packages/typehierarchy/src/browser/tree/typehierarchy-tree.ts

```diff
@@ -0,0 +1,200 @@
+ ********************************************************************************/
+
+import { inject, injectable } from 'inversify';
+import { v4 } from 'uuid';
+import URI from '@theia/core/lib/common/uri';
+import { Location } from '@theia/editor/lib/browser/editor';
+import { SymbolKind } from '@theia/languages/lib/browser';
+import { TreeDecoration } from '@theia/core/lib/browser/tree/tree-decorator';
+import { TreeImpl, TreeNode, CompositeTreeNode, ExpandableTreeNode, SelectableTreeNode } from '@theia/core/lib/browser/tree';
+import { TypeHierarchyItem } from '@theia/languages/lib/browser/typehierarchy/typehierarchy-protocol';
+import { TypeHierarchyType } from '@theia/languages/lib/browser/typehierarchy/typehierarchy-feature';
+import { TypeHierarchyService } from '../typehierarchy-service';
+
+@injectable()
+export class TypeHierarchyTree extends TreeImpl {
+
+    @inject(TypeHierarchyService)
+    protected readonly typeHierarchyService: TypeHierarchyService;
+
+    async resolveChildren(parent: CompositeTreeNode): Promise<TreeNode[]> {
+        if (TypeHierarchyTree.Node.is(parent)) {
+            await this.ensureResolved(parent);
+            if (parent.children.length === 0) {
+                delete parent.children;
+                delete parent.expanded;
+                return [];
+            }
+            return parent.children.slice();
+        }
+        return [];
+    }
+
+    /**
+     * Extracts the language ID from the root node.
+     */
+    protected get languageId(): string | undefined {
+        if (TypeHierarchyTree.RootNode.is(this.root)) {
+            return this.root.languageId;
+        }
+        return undefined;
+    }
+
+    /**
+     * Returns with the type hierarchy type attached to the root node. `undefined` if the root is not set.
+     */
+    protected get type(): TypeHierarchyType | undefined {
+        if (TypeHierarchyTree.RootNode.is(this.root)) {
+            return this.root.type;
+        }
+        return undefined;
+    }
+
+    /**
+     * Makes sure, the node and its children are resolved. Resolves it on demand.
+     */
+    protected async ensureResolved(node: TypeHierarchyTree.Node): Promise<void> {
+        if (!node.resolved) {
+            const { languageId, type } = this;
+            if (languageId && type) {
+                const { location } = node;
+                const resolvedItem = await (TypeHierarchyType.SUBTYPE === type
+                    ? this.typeHierarchyService.subTypes(languageId, location)
+                    : this.typeHierarchyService.superTypes(languageId, location));
+
+                if (resolvedItem) {
+                    node.resolved = true;
+                    const items = TypeHierarchyType.SUBTYPE === type ? resolvedItem.children : resolvedItem.parents;
+                    if (items) {
+                        node.children = items.map(child => TypeHierarchyTree.Node.create(child, type));
```

Comment:
Should this be passing `resolved = false` to `create()`? The protocol call above is using `resolve: 1`, so the node itself will be resolved but the children won't be.